core/trino-main/src/main/java/io/trino/operator/output/UnnestingPositionsAppender.java

```diff
@@ -31,10 +34,12 @@
     private static final int INSTANCE_SIZE = toIntExact(ClassLayout.parseClass(UnnestingPositionsAppender.class).instanceSize());
 
     private final PositionsAppender delegate;
+    private DictionaryBlockBuilder dictionaryBlockBuilder;
```

Comment:
Keep state top-level as in `io.trino.operator.output.RleAwarePositionsAppender`

Reply:
Since this class has now two responsibilities (unnesting and building a dictionary) it's more readable to separate them. It also makes it easier to reset the state of the appender.

Reply:
> Since this class has now two responsibilities (unnesting and building a dictionary)

This class doesn't really unnest much. `RleAwayPositionsAppender` could do:
```
        if (source instanceof RunLengthEncodedBlock) {
            delegate.appendRle(((RunLengthEncodedBlock) source).getValue(), positions.size());
        }
```
itself really at this point, so `UnnestingPositionsAppender` would be all about dictionaries

Reply:
`RleAwayPositionsAppender` is not always there. Unnesting part is actually to make sure only flat blocks are passed down to the flat appenders.
Maybe we should merge `UnnestingPositionsAppender` and `RleAwayPositionsAppender` into `BlockTypeAwarePositionsAppender`? although I fear it would make the code messier.

Reply:
> RleAwayPositionsAppender is not always there.

Why wouldn't it be always there?

Reply:
it's not needed if the type is not comparable

Reply:
> it's not needed if the type is not comparable

I don't think it's worth extra complexity since there are not many types like that.

However. You could then have minimal `UnnestingPositionsAppender` without rle or dict builder support.

I don't mixing of current `UnnestingPositionsAppender` with being dictionary aware is needed